src/boot/bootctl-uki.c

```diff
@@ -0,0 +1,96 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "bootctl.h"
+#include "bootctl-uki.h"
+#include "fd-util.h"
+#include "parse-util.h"
+#include "pe-header.h"
+
+#define DOS_FILE_MAGIC "MZ"
+#define PE_FILE_MAGIC  "PE\0\0"
+
+static const uint8_t name_osrel[8] = ".osrel";
+static const uint8_t name_linux[8] = ".linux";
+static const uint8_t name_initrd[8] = ".initrd";
+
+static struct PeSectionHeader* pe_sections(FILE *uki, uint16_t *n)
+{
+        struct PeSectionHeader *sections;
+        struct DosFileHeader dos;
+        struct PeHeader pe;
+        uint16_t scount;
+        off_t soff;
+        size_t rc;
+
+        rc = fread(&dos, 1, sizeof(dos), uki);
+        if (rc != sizeof(dos)) {
+                log_error("DOS header read error");
+                return NULL;
+        }
+        if (memcmp(dos.Magic, DOS_FILE_MAGIC, STRLEN(DOS_FILE_MAGIC)) != 0) {
+                log_error("DOS header magic mismatch");
+                return NULL;
+        }
+
+        fseek(uki, le32toh(dos.ExeHeader), SEEK_SET);
+        rc = fread(&pe, 1, sizeof(pe), uki);
+        if (rc != sizeof(pe)) {
+                log_error("PE header read error");
+                return NULL;
+        }
+        if (memcmp(pe.Magic, PE_FILE_MAGIC, STRLEN(PE_FILE_MAGIC)) != 0) {
+                log_error("PE header magic mismatch");
+                return NULL;
+        }
+
+        soff = le32toh(dos.ExeHeader) + sizeof(pe) + le16toh(pe.FileHeader.SizeOfOptionalHeader);
+        fseek(uki, soff, SEEK_SET);
+
+        scount = le16toh(pe.FileHeader.NumberOfSections);
+        sections = malloc(sizeof(*sections) * scount);
+        rc = fread(sections, sizeof(*sections), scount, uki);
+        if (rc != scount) {
+                log_error("PE section header read error");
+                free(sections);
+                return NULL;
+        }
+
+        *n = scount;
+        return sections;
```

Comment:
once switched to `_cleanup_free_` please use `return TAKE_PTR(sections);` here, to take the pointer out and invalidate it so that it isn't freed anymore.